### PR TITLE
SDI-113: ✨ Add in saving of mfa remember me token

### PR DESCRIPTION
**Pull request by petergphillips**
Screenshot from 2022-04-26 07-48-23

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserToken.kt [hunk 1]
```diff
@@ -52,8 +52,11 @@ class UserToken(
 
   fun resetExpiry() {
     val now = LocalDateTime.now()
-    tokenExpiry =
-      if (tokenType == TokenType.ACCOUNT || tokenType == TokenType.CHANGE || tokenType == TokenType.MFA) now.plusMinutes(20) else now.plusDays(1)
+    tokenExpiry = when (tokenType) {
+      TokenType.ACCOUNT, TokenType.CHANGE, TokenType.MFA -> now.plusMinutes(20)
+      TokenType.MFA_RMBR -> now.plusDays(7)
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserToken.kt:57]: MFA remember me token valid for 7 days

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaServiceBasedController.kt [hunk 2]
```diff
@@ -54,21 +63,29 @@ class MfaServiceBasedController(
     @RequestParam token: String?,
     @RequestParam mfaPreference: MfaPreferenceType?,
     @RequestParam user_oauth_approval: String?,
-  ): ModelAndView = mfaChallengeRequest(error, token, mfaPreference, extraModel(user_oauth_approval))
+    @ModelAttribute("authorizationRequest") authorizationRequest: AuthorizationRequest,
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaServiceBasedController.kt:66]: the `authorizationRequest` is stored in the session so is already available

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaClientService.kt [hunk 3]
```diff
@@ -16,16 +16,14 @@ open class MfaClientService(
     val client = clientDetailsService.loadClientByClientId(request?.clientId)
     val baseClientId = ClientService.baseClientId(client.clientId)
 
+    // Special case for migration of check my diary users - ignore 2fa if they don't have the migrated role
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/MfaClientService.kt:19]: comment in wrong place from previous PR

File: src/main/resources/db/dev/data/auth-postgresql/V900_0__clients.sql [hunk 4]
```diff
@@ -84,7 +84,8 @@ VALUES ('another-delete-test-client', '1200', '{}', null, 'client_credentials',
...
        ('service-add-test-client', 1200, '{"jwtFields":"-name"}', null, 'authorization_code', 'read', '$2a$10$lBwbziQlLfiCnn8Kj1PfMujEcLdsJYlYSNJvBRO638gCYTS9yN0xm', null, null, 'read', 'http://localhost:3001/,http://localhost:3001/sign-in/callback'),
        ('service-edit-test-client', 1200, '{"jwtFields":"-name"}', null, 'authorization_code', 'read', '$2a$10$lBwbziQlLfiCnn8Kj1PfMujEcLdsJYlYSNJvBRO638gCYTS9yN0xm', null, null, 'read', 'http://localhost:3001/,http://localhost:3001/sign-in/callback'),
-       ('service-mfa-test-client','3600','{"mfa":"all","mfaRememberMe":true}',null,'authorization_code',null,'$2a$10$lBwbziQlLfiCnn8Kj1PfMujEcLdsJYlYSNJvBRO638gCYTS9yN0xm',null,null,'read','http://localhost:8081/login'),
+       ('service-mfa-test-client','3600','{"mfa":"all"}',null,'authorization_code',null,'$2a$10$lBwbziQlLfiCnn8Kj1PfMujEcLdsJYlYSNJvBRO638gCYTS9yN0xm',null,null,'read','http://localhost:8081/login'),
+       ('service-mfa-remember-test-client','3600','{"mfa":"all","mfaRememberMe":true}',null,'authorization_code',null,'$2a$10$lBwbziQlLfiCnn8Kj1PfMujEcLdsJYlYSNJvBRO638gCYTS9yN0xm',null,null,'read','http://localhost:8081/login'),
```
Comment (petergphillips) [inline src/main/resources/db/dev/data/auth-postgresql/V900_0__clients.sql:88]: created new client for testing the remember me

File: src/main/resources/templates/mfaChallengeServiceBased.html [hunk 5]
```diff
@@ -64,8 +64,15 @@ <h1 class="govuk-heading-l" th:unless="${#strings.equals(mfaPreference, 'TEXT')}
             id="code" name="code" type="text" inputmode=”numeric” pattern="[0-9]*" autocomplete="off"
             autocapitalize="off" spellcheck="false" data-element-id="code" aria-describedby="code-error" maxlength="6">
       </div>
+    </div>
 
+    <div class="govuk-form-group" th:if="${mfaRememberMe}">
```
Comment (petergphillips) [inline src/main/resources/templates/mfaChallengeServiceBased.html:69]: only shown if client remember me set

**Review by SimonMitchellMOJ — APPROVED** [review]
LGTM
